### PR TITLE
[15.0][MIG] sale_triple_discount

### PR DESCRIPTION
**Pull request by leemannd**
Depends on:

- [x] https://github.com/OCA/account-invoicing/pull/1144

### CODE REVIEW

**Review by SilvioC2C — APPROVED** [review]
LGTM

**Review by ajaniszewska-dev — APPROVED** [review]

File: sale_triple_discount/models/sale_order_line.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            return self._multiplicative_discount()
+        else:
+            raise ValidationError(
+                _("Sale order line %(name)s has unknown discounting type %(dic_type)s")
+                % {"name": self.name, "disc_type": self.discounting_type}
+            )
+
+    def _additive_discount(self):
+        self.ensure_one()
+        discount = sum(getattr(self, x) or 0.0 for x in self._discount_fields())
```
Comment (simahawk) [inline sale_triple_discount/models/sale_order_line.py:28]:
```suggestion
        discount = sum(self[x] or 0.0 for x in self._discount_fields())
```

File: sale_triple_discount/models/sale_order_line.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        return ["discount", "discount2", "discount3"]
+
+    @api.depends("discount2", "discount3", "discounting_type")
+    def _compute_amount(self):
+        prev_values = self.triple_discount_preprocess()
+        res = super()._compute_amount()
+        self.triple_discount_postprocess(prev_values)
+        return res
+
+    discount2 = fields.Float(
```
Comment (simahawk) [inline sale_triple_discount/models/sale_order_line.py:53]: can we move these fields up?

File: sale_triple_discount/models/sale_order_line.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            "CHECK (discount3 <= 100.0)",
+            "Discount 3 must be lower than 100%.",
+        ),
+    ]
+
+    def _get_triple_discount(self):
+        """Get the discount that is equivalent to the subsequent application
+        of discount, discount2 and discount3"""
+        discount_factor = 1.0
+        for discount in [self.discount, self.discount2, self.discount3]:
```
Comment (simahawk) [inline sale_triple_discount/models/sale_order_line.py:90]:
```suggestion
        for discount in (self.discount, self.discount2, self.discount3):
```

File: sale_triple_discount/models/sale_order_line.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        res = super()._get_price_reduce()
+        self.triple_discount_postprocess(prev_values)
+        return res
+
+    def triple_discount_preprocess(self):
+        """Save the values of the discounts in a dictionary,
+        to be restored in postprocess.
+        Resetting discount2 and discount3 to 0.0 avoids issues if
+        this method is called multiple times.
+        Updating the cache provides consistency through recomputations."""
```
Comment (simahawk) [inline sale_triple_discount/models/sale_order_line.py:111]:
```suggestion
        """Prepare data for post processing.
        
        Save the values of the discounts in a dictionary,
        to be restored in postprocess.
        Resetting discount2 and discount3 to 0.0 avoids issues if
        this method is called multiple times.
        Updating the cache provides consistency through recomputations."""
```

File: sale_triple_discount/models/sale_order_line.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+                    "discount2": 0.0,
+                    "discount3": 0.0,
+                }
+            )
+        return prev_values
+
+    @api.model
+    def triple_discount_postprocess(self, prev_values):
+        """Restore the discounts of the lines in the dictionary prev_values.
+        Updating the cache provides consistency through recomputations."""
```
Comment (simahawk) [inline sale_triple_discount/models/sale_order_line.py:134]:
```suggestion
        """Restore the discounts of the lines in the dictionary prev_values.

        Updating the cache provides consistency through recomputations."""
```

File: sale_triple_discount/tests/test_sale_triple_discount.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo.tests import common
+
+
+class TestSaleOrder(common.SavepointCase):
+    @classmethod
+    def setUpClass(cls):
+        super(TestSaleOrder, cls).setUpClass()
+        cls.partner = cls.env["res.partner"].create({"name": "Mr. Odoo"})
```
Comment (simahawk) [inline sale_triple_discount/tests/test_sale_triple_discount.py:12]:
```suggestion
        cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
        cls.partner = cls.env["res.partner"].create({"name": "Mr. Odoo"})
```

**Comment by rescola** [comment]
image
seems there's problem with vat, it shows 18,9 vat, but 72,9 * 0,21 = 15,31, but the total is correct 72,9 + 15,31 = 88,21.
Then is a problem showing taxes.

Also fount problem selecting additive discount, then creating it as an invoice it transform it again to multiplicative, account_invoice_triple_discount doesn't have this "type" and considers all invoices as multiplicative:
This the sale order:
image
and the invoice (don't match)
image

**Comment by ValentinVinagre** [comment]
@leemannd do you plan to finish the pr?

**Pull request closed without merging**